### DISCUSSION TITLE
Restore old crop feature from v3.16 or fix new one?

### DISCUSSION THREAD

**Discussion by c0xc** [c0xc/nomacs#7] [General]
I want to make a couple of improvements in Nomacs but I'm trying to figure out which base version I should use. The latest dev version is v3.17 which has a new crop feature which isn't working right (see #4).

Option 1: Fork v3.16 and continue working from there, restoring the old crop feature which was working fine.
Option 2: Try to fix this new crop feature for Nomacs to become usable again.

In the old version you would select a section of the image, which you were able to zoom to make a precise cut. In the new version, it basically works the other way around, so you'd have a white grid centered in the middle of the window and instead of moving that section, you are moving the full image in the background. If you try to zoom out, the grid moves. Sometimes the grid jumps around, I haven't yet figured out when and how, exactly. All I can say is that cropping an image, something which used to be so simple that I never really thought about it, is now very difficult or, if the selection keeps jumping around, impossible.

I'm hesitant to remove a new feature, something someone put a lot of effort in. But in my view, it's not ready for production at all. And to be honest, I don't exactly get why this feature has been rewritten, why you're now supposed to move the background image instead of the section you're after. I suppose it could be fixed but it still seems counter-intuitive, definitely not as easy as before...

**Comment by novomesk** [comment]
Some info about original idea: https://github.com/nomacs/nomacs/issues/345 ?

**Reply by c0xc** [reply]
It sounds like the idea was to make this feature behave like Photoshop. I'm not a PS user, how would you crop an area that doesn't fit in your window (and you have to zoom to carefully select the area)? The video in this comment shows the issue:
https://github.com/nomacs/nomacs/issues/345#issuecomment-671794311

**Reply by novomesk** [reply]
At least in old Photoshop, the crop rectangle zooms together with image.

**Reply by c0xc** [reply]
In Nomacs v3.16, the crop rectangle also zoomed together with the image. You can easily select the rough area you're interested in, then zoom into the top left corner to make minor adjustments to your selection and then zoom out and zoom into the bottom right corner to make sure it's also where you want to have it. But in v3.17 none of that works because the crop rectangle has a fixed size, relative to the size of the window. You cannot zoom in to make an exact selection, as zooming changes your selection. To be honest, this new crop in v3.17 is so frustrating, I'd rather just fork and keep working with the old crop which worked just fine. What do you think? (The only thing I didn't like about the old one was that everytime you clicked outside of the selected area, you'd rotate the selected area. I don't get why a rotate feature was mixed into this crop feature but even that worked much better than v3.17 which rotates up to 45 degrees, not more.)

**Reply by novomesk** [reply]
I think it is reasonable to restore the old crop.
But there are also many good commits after 3.16

**Reply by c0xc** [reply]
I agree on both points. That's probably going to be a lot of work. As of now, I'm personally interested in fixing Nomacs because I use it myself. However, my time is limited. It seems like I'll have to dig through the changes since v3.16 and identify those that I want to keep and that'll take a while...

**Reply by novomesk** [reply]
I think if you identify all commits related to the new crop feature, it should be possible to do interactive git rebase and select only other commits.

**Reply by novomesk** [reply]
I have now permission https://github.com/nomacs/nomacs, do you think you will be able to fix the "new crop" from 3.17?

**Reply by c0xc** [reply]
Sounds good! Give me some time. I've been playing around with it, allowing the background to be moved exclusively (only when the mouse cursor is on the background), but trying to stop the jumping around would require a lot of debugging of things that used to work just fine. I thought I had found a good starting point (correction, I had found a fairly recent commit before all of the behavior changes but it turns out that commit wasn't working) but I'm reluctant to do a rebase as that would break currently checked out copies of the repo and future merges. I think I'll try to selectively revert those changes...

**Reply by c0xc** [reply]
I have a test version in "pse/crop_revert". I will add the latest changes and push that into a new staging branch, probably "integration" (so we have more time to test it before it goes into master, same for other people who want to test it).

**Comment by clownfishnemoGitHub** [comment]
> I want to make a couple of improvements in Nomacs but I'm trying to figure out which base version I should use.
I also wanted to create a fork, to try out to get the build going and try and test things out
i am very new to this github related but have a fair decent amount of knowledge for programming using c language

which base should i use to create a nomacs fork from ?

at nomacs i find out it is not possible, to launch and open a folder path to a mobile phone which is hooked up
with usb cable to a laptop, i won't take or allow that as folder path to be used to browse images on a folder directly on
my phone camera gallery images folder, this will be my main focus as new enhancement for nomacs

cheers
